akka-stream/src/main/scala/akka/stream/stage/Stage.scala

```diff
@@ -27,10 +28,68 @@ import akka.stream.Supervision
  */
 sealed trait Stage[-In, Out]
 
-private[stream] abstract class AbstractStage[-In, Out, PushD <: Directive, PullD <: Directive, Ctx <: Context[Out]] extends Stage[In, Out] {
-  private[stream] var holding = false
-  private[stream] var allowedToPush = false
-  private[stream] var terminationPending = false
+/**
+ * INTERNAL API
+ */
+private[stream] object AbstractStage {
+  final val UpstreamBall = 1
+  final val DownstreamBall = 2
+  final val BothBalls = UpstreamBall | DownstreamBall
+  final val PrecedingWasPull = 0x4000
+  final val TerminationPending = 0x8000
+}
+
+abstract class AbstractStage[-In, Out, PushD <: Directive, PullD <: Directive, Ctx <: Context[Out]] extends Stage[In, Out] {
+  /**
+   * INTERNAL API
+   */
+  private[stream] var bits = 0
+
+  /**
+   * INTERNAL API
+   */
+  private[stream] var context: Ctx = _
+
+  /**
+   * INTERNAL API
+   */
+  private[stream] def isDetached: Boolean = false
+
+  /**
+   * INTERNAL API
+   */
+  private[stream] def enterAndPush(elem: Out): Unit = {
+    context.enter()
+    context.push(elem)
+    context.execute()
+  }
+
+  /**
+   * INTERNAL API
+   */
+  private[stream] def enterAndPull(): Unit = {
```

Comment:
can these be pushed down to subclasses that actually use them? Only BoundaryStage and AsyncStage uses these.